src/kirin/dialects/py/len.py

```diff
@@ -55,3 +57,50 @@ def lower_Call_len(
         return lowering.Result(
             state.append_stmt(Len(value=state.visit(node.args[0]).expect_one()))
         )
+
+
+class InferLen(RewriteRule):
+    EllipsisType = types.PyClass(type(...))
+
+    @cached_property
+    def Constant(self):
+        # Avoid circular import caching the result
+        from kirin.dialects.py.constant import Constant
+
+        return Constant
+
+    @cached_property
+    def IListType(self):
+        # Avoid circular import caching the result
+        from kirin.dialects.ilist import IListType
+
+        return IListType
+
+    def _get_collection_len(self, collection: ir.SSAValue):
+        coll_type = collection.type
+
+        if not isinstance(coll_type, types.Generic):
+            return None
+
+        if (
+            coll_type.is_subseteq(self.IListType)
+            and isinstance(coll_type.vars[1], types.Literal)
+            and isinstance(coll_type.vars[1].data, int)
+        ):
+            return coll_type.vars[1].data
+        elif coll_type.is_subseteq(types.Tuple) and not any(
+            var.is_subseteq(self.EllipsisType) for var in coll_type.vars
+        ):
+            return len(coll_type.vars)
```

Comment:
this is wrong, why do we need to deal with `tuple` at all inside `ilist` dialect? Isn't this case being covered by constprop?

Reply:
and you are mixing up Python types with Kirin types here... You want `types.Vararg` not `...`

Reply:
I want to make sure none of the types inside the tuple isn't `Ellipses` , e.g., `Tuple[int, ...]` because that doesn't have a well defined length even though it has two parameters.

Reply:
> this is wrong, why do we need to deal with `tuple` at all inside `ilist` dialect? Isn't this case being covered by constprop?

I'm handing these cases separately. I'm not handling nested types. Constant prop doesn't always run because of missing concrete interpreters but we can get this from type inference.

Reply:
Constant prop uses a PartialTuple lattice element to keep track of tuples. This includes the length of a tuple despite its value.

Reply:
You are dealing with Kirin's type system now so you don't have ellipsis there's only Vararg type in generics.